text/0088-metric-instrument-optional-refinements.md

```diff
@@ -0,0 +1,483 @@
+here, in the sense that any kind of metric instrument must reduce to
+one of these.  The foundational instruments are unrestricted, in the
+sense that metric events support any numerical value, positive or
+negative, zero or infinity.
+
+The distinction between the two foundational instruments is whether
+they are synchronous.  Measure instruments are called synchronously by
+the user, while Observer instruments are called asynchronously by the
+implementation.  Synchronous instruments (Measure and its refinements)
+have three calling patterns (_Bound_, _Unbound_, and _Batch_) to
+capture measurements.  Asynchronous instruments (Observer and its
+refinements) use callbacks to capture measurements.
+
+All measurement APIs produce metric events consisting of [timestamp,
+instrument descriptor, label set, and numerical
+value](api-metrics.md#metric-event-format).  Synchronous instrument
+events additionally have [Context](api-context.md), describing
+properties of the associated trace and distributed correlation values.
+
+### Last-value relationship
+
+Observer instruments have a well-defined _Last Value_ measured by the
+instrument, that can be useful in defining aggregations.  To maintain
+this property, we impose this requirement: two or more `Observe()`
+calls with an identical LabelSet during a single Observer callback
+invocation are treated as duplicates of each other, where the last
+call to `Observe()` wins.  (This is also the specified way that
+`Labels()` treats duplicates for a given label key.)
+
+Unlike Observer instruments, Measure instruments do not define the
+Last Value relationship.  One reason is that [synchronous events can
+happen
+simultaneously](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics.md#time).
+The Last Value relationship refers to values read in a single
+collection period, whereas with Measure instruments we can define a
+last-value aggregation.
+
+### Aggregating changes to a sum: Rate calculation
+
+The former `Monotonic` option had been introduced in order to support
+reporting of a current sum, such that a rate calculation is implied.
+Here we defined _Rate_ as an aggregation, defined for a subset of
+instruments, that may be calculated differently depending on how the
+instrument is defined.  The rate aggregation outputs the amount of
+change in a quantity divided by the amount of change in time.
+
+A rate can be computed from values that are reported as differences,
+referred to as _delta_ reporting here, or as sums, referred to as
+_cumulative_ reporting here.  The primary goal of the instrument
+refinements introduced in this proposal is to facilitate rate
+calculations in more than one way.
+
+When delta reporting, a rate is calculated by summing individual
+measurements or observations.  For Measure instruments, these values
+fall into a range of time, as indicated by the event timestamp.  For
```

Comment:
Does this mean that for `Measure` and `Counter` instruments, the `timestamp` attribute represents the time of recording, and for `Observer` it represents the time of collection? Since observers are recorded at time of collection, is it safe to say that all `timestamp` values are timestamps for each event when recorded. Does this go against what [the specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics.md#time) say in regards to timestamp?

Reply:
I don't see a disagreement there. In the linked specification section, there's an effort to say that timestamps are implicit and might not actually be captured. So an SDK is free to use a single timestamp for all events in an interval, if it think's that is reasonable. For Observer instruments, I was trying to state that all observations for an interval happen at logically the same time. They are logically recorded at the same time, which I think satisfies the statement you're making.
(I'm not sure I'm helping here.)

Reply:
I guess the key here "...free to..."? I can see use cases where an explicit timestamp is required for certain ingestion services (Azure for example), so if there is flexibility in how the SDK implements it that would be quite versatile.

Reply:
I take this as a recommendation that future versions of the default SDK should support different options with regards to timestamps.

How would you treat a Counter instrument, for example? If there are many `Add()` calls throughout a collection interval, the standard behavior is to add them without computing timestamps for each. Even the `Exact` aggregator, which simply builds an array of values, does not bother to store the timestamp for each. Only the current "lastvalue" aggregator bothers with timestamps, which it uses to maintain the last value. I can imagine an `ExactWithTimestamp` aggregator that would build an array of (value, timestamp) pairs. That could be used, but I was under the impression that even the Azure system was implemented by using 1-second checkpoints with a single timestamp for the whole 1-second batch.